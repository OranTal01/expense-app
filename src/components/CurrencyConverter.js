import React, { Component } from "react";

class CurrencyConverter extends React.Component {
    state = {currencies:["CAD","HKD","ISK","PHP","DKK","HUF","CZK","GBP","RON","SEK","IDR","INR","BRL","RUB","HRK","JPY",
"THB","CHF","EUR","MYR","BGN","TRY","CNY","NOK","NZD","ZAR","USD","MXN","SGD","AUD","ILS","KRW","PLN"],
        base: "USD",
        amount: "",
        convertTo: "ILS",
        result: "",
        date: "",
        error: "",
        amountConverter: ""
    };
    onCurrencyConverter = (e) => {
        e.preventDefault();
        const amount = this.state.amount;
        if (!amount) {
            this.setState(() => ({
                error: 'Please provide amount to converter.'
            }))
        } else {
            fetch(`https://api.exchangeratesapi.io/latest?base=${this.state.base}`)
            .then(res => res.json())
            .then(data => {
                const date = data.date;
                const result = (data.rates[this.state.convertTo] * amount).toFixed(2);
                const amountConverter = data.rates[this.state.convertTo];
                this.setState({
                    date,
                    result,
                    amountConverter
                });
            });
        }
    };
    
    handleInput = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }
    onBaseCurrencyChange = (e) => {
        const base = e.target.value;
        this.setState(()=>({base}))
    };
    onConvertCurrencyChange = (e) => {
        const convertTo = e.target.value;
        this.setState(()=>({convertTo}))
    };

    onSwapCurrency = () => {
        const tempBase = this.state.base;
        const tempConvertTo = this.state.convertTo;
        this.setState(() => ({
            base: tempConvertTo,
            convertTo: tempBase,
            amount: "",
            result: "",
            date: "",
            amountConverter: ""
        }));
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onCurrencyConverter}>
                    <div>
                    Base Currency:
                    <input type="text"
                            value={ this.state.amount }
                            onChange={ this.handleInput}/>
                        <select
                        value={ this.state.base }
                        onChange={this.onBaseCurrencyChange}>
                            <option value="USD">USD</option>
                            <option value="EUR">EUR</option>
                            <option value="ILS">ILS</option>
                            <option value="GBP">GBP</option>
                            <option value="CAD">CAD</option>
                            <option value="HKD">HKD</option>
                            <option value="ISK">ISK</option>
                            <option value="PHP">PHP</option>
                            <option value="DKK">DKK</option>
                            <option value="HUF">HUF</option>
                            <option value="CZK">CZK</option>
                            <option value="RON">RON</option>
                            <option value="SEK">SEK</option>
                            <option value="IDR">IDR</option>
                            <option value="INR">INR</option>
                            <option value="BRL">BRL</option>
                            <option value="RUB">RUB</option>
                            <option value="HRK">HRK</option>
                            <option value="JPY">JPY</option>
                            <option value="THB">THB</option>
                            <option value="CHF">CHF</option>
                            <option value="MYR">MYR</option>
                            <option value="BGN">BGN</option>
                            <option value="TRY">TRY</option>
                            <option value="CNY">CNY</option>
                            <option value="NOK">NOK</option>
                            <option value="NZD">NZD</option>
                            <option value="ZAR">ZAR</option>
                            <option value="MXN">MXN</option>
                            <option value="SGD">SGD</option>
                            <option value="AUD">AUD</option>
                            <option value="KRW">KRW</option>
                            <option value="PLN">PLN</option>
                    </select>
                    </div>
                    <div>
                    Convert To:
                        <select
                        value={ this.state.convertTo }
                        onChange={this.onConvertCurrencyChange}>
                        <option value="USD">USD</option>
                        <option value="EUR">EUR</option>
                        <option value="ILS">ILS</option>
                        <option value="GBP">GBP</option>
                        <option value="CAD">CAD</option>
                        <option value="HKD">HKD</option>
                        <option value="ISK">ISK</option>
                        <option value="PHP">PHP</option>
                        <option value="DKK">DKK</option>
                        <option value="HUF">HUF</option>
                        <option value="CZK">CZK</option>
                        <option value="RON">RON</option>
                        <option value="SEK">SEK</option>
                        <option value="IDR">IDR</option>
                        <option value="INR">INR</option>
                        <option value="BRL">BRL</option>
                        <option value="RUB">RUB</option>
                        <option value="HRK">HRK</option>
                        <option value="JPY">JPY</option>
                        <option value="THB">THB</option>
                        <option value="CHF">CHF</option>
                        <option value="MYR">MYR</option>
                        <option value="BGN">BGN</option>
                        <option value="TRY">TRY</option>
                        <option value="CNY">CNY</option>
                        <option value="NOK">NOK</option>
                        <option value="NZD">NZD</option>
                        <option value="ZAR">ZAR</option>
                        <option value="MXN">MXN</option>
                        <option value="SGD">SGD</option>
                        <option value="AUD">AUD</option>
                        <option value="KRW">KRW</option>
                        <option value="PLN">PLN</option>
                        </select>
                        { this.state.amountConverter &&
                            <p>your Currency Converter is:
                        { this.state.amountConverter }</p> }
                    </div>
                    <button>
                        Start Converter
                    </button>
                </form>
                <button onClick={ this.onSwapCurrency }>
                Swap currency
                </button>
                { this.state.date && <p>Date of Currency Converter: { this.state.date }</p>}
                { this.state.result && <p>Convert To: { this.state.result } { this.state.convertTo }</p>}
            </div>
        )
    }
}

export default CurrencyConverter;