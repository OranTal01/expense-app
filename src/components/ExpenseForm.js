import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import {history} from '../routers/AppRouter';


export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      payments: '1'
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    }
    else if (this.state.payments > 1) {
      const userPayments = this.state.payments
      this.setState(() => ({ error: '' }));
      for (let index = userPayments; index >= 1; index--) {
        this.props.onSubmit({
          description: this.state.description,
          amount: (parseFloat(this.state.amount, 10) * 100) / (parseFloat(userPayments,10)),
          createdAt: index === userPayments ? this.state.createdAt.valueOf() : this.state.createdAt.add(1, 'months').valueOf(),
          note: this.state.note
        });
      }
    }
    else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  toggleAddEditExpense = () => {
    if (history.location.pathname === '/create') {
      return true
    } else {
      return false
    }
  };

  onChangePayments = (e) => {
    const userPayments = e.target.value;
    this.setState(() => ({
      payments: userPayments
    }));
  };

  render() {
    return (
      <form
      className="form"
        onSubmit={ this.onSubmit }>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
      <input
      className="text-input"
      type="text"
      placeholder="Description"
      autoFocus
      value={this.state.description}
      onChange={this.onDescriptionChange}
  />
  <input
    className="text-input"
    type="text"
    placeholder="Amount"
    value={this.state.amount}
    onChange={this.onAmountChange}
  />
  <SingleDatePicker
    date={this.state.createdAt}
    onDateChange={this.onDateChange}
    focused={this.state.calendarFocused}
    onFocusChange={this.onFocusChange}
    numberOfMonths={1}
    isOutsideRange={() => false}
  />
  <textarea
    className="textarea"
    placeholder="Add a note for your expense (optional)"
    value={this.state.note}
    onChange={this.onNoteChange}
  >
  </textarea>
  <div>
    <select
        className="select"
        value={this.state.payments}
        onChange={ this.onChangePayments }>
        <option value="1">Payments: 1</option>
        <option value="2">Payments: 2</option>
        <option value="3">Payments: 3</option>
        <option value="4">Payments: 4</option>
        <option value="5">Payments: 5</option>
        <option value="6">Payments: 6</option>
        <option value="7">Payments: 7</option>
        <option value="8">Payments: 8</option>
        <option value="9">Payments: 9</option>
        <option value="10">Payments: 10</option>
        <option value="11">Payments: 11</option>
        <option value="12">Payments: 12</option>
        <option value="13">Payments: 13</option>
        <option value="14">Payments: 14</option>
        <option value="15">Payments: 15</option>
        <option value="16">Payments: 16</option>
        <option value="17">Payments: 17</option>
        <option value="18">Payments: 18</option>
        <option value="19">Payments: 19</option>
        <option value="20">Payments: 20</option>
        <option value="21">Payments: 21</option>          
        <option value="22">Payments: 22</option>
        <option value="23">Payments: 23</option>
        <option value="24">Payments: 24</option>
      </select>
      </div>
      <div>
        { this.toggleAddEditExpense()
        ?
        <button className="button">Save Expense</button>
        :
        <button className="button">Save Edit Expense</button> }
      </div>
      </form>      
    )
  }
}
