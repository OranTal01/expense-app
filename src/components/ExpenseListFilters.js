import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';


class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            focusedInput: null
        }
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }
    onFocusChange = (focusedInput) => {
        this.setState(() => ({ focusedInput }))
    }

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={ this.props.filters.text }
                    onChange={ (e) => {
                        this.props.dispatch(setTextFilter(e.target.value));
                    } }
                />
                <select
                    value={ this.props.filters.sortBy }
                    onChange={ (e) => {
                        if (e.target.value === 'date') {
                            this.props.dispatch(sortByDate());
                        } else if (e.target.value === 'amount') {
                            this.props.dispatch(sortByAmount());
                        }
                    } }
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDateId="MyDatePickerStart"
                    endDateId="MyDatePickerEnd"
                    startDate={ this.props.filters.startDate } // momentPropTypes.momentObj or null,
                    endDate={ this.props.filters.endDate } // momentPropTypes.momentObj or null,
                    onDatesChange={ this.onDatesChange }// PropTypes.func.isRequired,
                    focusedInput={ this.state.focusedInput } // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={ this.onFocusChange } // PropTypes.func.isRequired,
                    numberOfMonths={ 1 }
                    isOutsideRange={ () => false }
                    showClearDates={ true }
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);
