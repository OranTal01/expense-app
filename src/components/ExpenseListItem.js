import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  if (numeral.locales["user-locale"] === undefined) {
    numeral.register("locale", "user-locale", {
      delimiters: {
        thousands: ",",
        decimal: "."
      },
      abbreviations: {
        thousand: "k",
        million: "m",
        billion: "b",
        trillion: "t"
      },
      currency: {
        symbol: "â‚ª"
      }
    });
    numeral.locale("user-locale");
  }

  const amountShekel = numeral(amount / 100).format("$0,0.00");
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        {amountShekel}-{moment(createdAt).format("DD MMMM, YYYY")}
      </p>
    </div>
  );
};

export default ExpenseListItem;