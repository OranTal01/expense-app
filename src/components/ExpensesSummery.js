import React from 'react';
import selectExpensesTotal  from '../selectors/expenses-total';
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses';
import numeral from "numeral";

if (numeral.locales["user-locale"] === undefined) {
    numeral.register("locale", "user-locale", {
        delimiters: {
        thousands: ",",
        decimal: "."
        },
        abbreviations: {
        thousand: "k",
        million: "m",
        billion: "b",
        trillion: "t"
        },
        currency: {
        symbol: "â‚ª"
        }
    });
    numeral.locale("user-locale");
    }

export const ExpensesSummery = (props) => {
    const expenseTotal = selectExpensesTotal(props.expenses);
    const convertExpenseTotal = numeral(expenseTotal / 100).format("$0,0.00");
    const expenseAmount = props.expenses.length
    return (
        <div>
            { props.expenses.length > 0 ? <p> Viewing { expenseAmount } expenses totalling
            { convertExpenseTotal }</p> : '' }
        </div>
    )
};

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpensesSummery);
