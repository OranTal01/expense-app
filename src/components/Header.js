import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import {firebase} from '../firebase/firebase';

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <h3 className="header__title--user">
        <img className="user-icon" width="30" height="30" src={firebase.auth().currentUser.photoURL} />
          { firebase.auth().currentUser.displayName }
        </h3>
    <Link className="header__title" to="/dashboard">
      <h1>Expensify</h1>
    </Link>
    <Link className="header__title" to="/Currency">
      <h1>Currency Converter</h1>
      </Link>
      <button className="button button--link" onClick={startLogout}>Logout</button>
  </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined,mapDispatchToProps)(Header);
