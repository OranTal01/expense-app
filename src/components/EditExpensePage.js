import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import RemoveExpenseModal from './RemoveExpenseModal';

export class EditExpensePage extends React.Component {
  state = {
    removeExpenseModal: undefined
  }
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  handelRemoveExpense = () => {
    this.setState(() => ({
      removeExpenseModal: undefined
    }));
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  onOpenRemoveModal = () => {
    this.setState(() => ({
      removeExpenseModal : this.props.expense.description
    }));
  };

  handelCloseModal = () => {
    this.setState(() => ({
      removeExpenseModal: undefined
    }));
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1
              className="page-header__title">
              Edit Expense
            </h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          />
          <div>
            <RemoveExpenseModal
              openRemoveExpenseModal={ this.state.removeExpenseModal }
              handelRemoveExpense={ this.handelRemoveExpense }
              handelCloseModal={ this.handelCloseModal}/>
          <button
            className="button button--secondary"
            onClick={ this.onOpenRemoveModal }>
            Remove Expense
          </button>
          </div>
          </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
