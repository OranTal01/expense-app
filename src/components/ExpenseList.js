import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {
      props.expenses.length === 0 ? (<p>No Expenses</p>) :
        (props.expenses.map((expense) => {
          return <ExpenseListItem key={ expense.id } { ...expense } />
        }))
    }
    <h3>Total Amount:{ props.expenses.reduce((a, b) => parseFloat(a) + parseFloat(b.amount), 0).toFixed(2) }</h3>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);