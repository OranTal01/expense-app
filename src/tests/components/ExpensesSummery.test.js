import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummery} from '../../components/ExpensesSummery';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummery with expenses', () => {
    const wrapper = shallow(<ExpensesSummery expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});
    