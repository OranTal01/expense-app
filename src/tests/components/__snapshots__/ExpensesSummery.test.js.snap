// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummery with expenses 1`] = `
<div>
  <p>
     Viewing 
    3
     expenses totalling
    â‚ª1,141.95
  </p>
</div>
`;
